src/main/java/org/passay/HaveIBeenPwnedRule.java

```diff
@@ -0,0 +1,227 @@
+  }
+
+
+  /**
+   * maximum Duration for reading from the API.
+   *
+   * @param timeout for reading.
+   */
+  public void setReadTimeout(final Duration timeout)
+  {
+    readTimeout = timeout;
+  }
+
+
+  /**
+   * If an exception occurs during accessing the api, the password will be allowed, if set to true.
+   *
+   * @param allow true: if the API is not accessible, any password is accepted.
+   *              false: Default, API must answer in time to allow the password.
+   */
+  public void setAllowOnException(final boolean allow)
+  {
+    allowOnException = allow;
+  }
+
+
+  @Override
+  public RuleResult validate(final PasswordData passwordData)
+  {
+    final String hexDigest = getHexDigest(passwordData);
+    try (LineNumberReader lnr = openApiConnectionForRange(hexDigest.substring(0, PREFIX_LENGTH))) {
+      return searchResponse(hexDigest, lnr);
+    } catch (IOException e) {
+      return new RuleResult(allowOnException,
+        new RuleResultDetail(IO_ERROR_CODE, Collections.singletonMap("url", apiUrl)));
+    }
+  }
+
+
+  /**
+   * Reads the supplied reader line by line until a match is found against the supplied hex digest.
+   *
+   * @param hexDigest to match
+   * @param reader to read
+   *
+   * @return rule result whose validity is determined by whether a match was found
+   *
+   * @throws IOException if an error occurs reading from the reader
+   */
+  private RuleResult searchResponse(final String hexDigest, final LineNumberReader reader) throws IOException
+  {
+    String line;
+    final Pattern p = Pattern.compile("^(" + hexDigest.substring(PREFIX_LENGTH) + "):(\\d+)\\s*$");
+    while ((line = reader.readLine()) != null) {
+      final Matcher m = p.matcher(line);
+      if (m.matches()) {
+        final int matchCount = Integer.parseInt(m.group(2));
+        return new RuleResult(allowExposed,
+          new RuleResultDetail(ERROR_CODE, Collections.singletonMap("count", matchCount)),
+          new RuleResultMetadata(RuleResultMetadata.CountCategory.Pwned, matchCount));
+      }
+    }
+    return new RuleResult(true);
+  }
+
+
+  /**
+   * Returns an uppercase, hex encoded, SHA1 hash of the password.
+   *
+   * @param passwordData to hash
+   *
+   * @return hex encoded hash
+   */
+  private static String getHexDigest(final PasswordData passwordData)
+  {
+    final byte[] digest = HashUtil.sha1(passwordData.getPassword().getBytes(Charset.defaultCharset()));
+    return CodecUtil.encode(new HexEncoder(false, true), digest);
+  }
+
+
+  /**
+   * Opens a connection to the API and returns a reader for the input stream.
+   *
+   * @param range to request from the API
+   *
+   * @return reader for the connection
+   *
+   * @throws IOException if a connection cannot be opened to the API
+   */
+  private LineNumberReader openApiConnectionForRange(final String range)
+    throws IOException
+  {
+    final URL url = new URL(apiUrl, range);
+    final URLConnection c = url.openConnection();
```

Comment:
May want to consider HTTP connection pooling here. That said I appreciate the simplicity of this code using JDK libraries.

Reply:
Seems like a v2 feature. Note that persistent connections are supported:
https://docs.oracle.com/javase/8/docs/technotes/guides/net/http-keepalive.html